internal/edittree/edittree.go

```diff
@@ -0,0 +1,1135 @@
+			}
+			// Normal value case.
+			subtreeResult := child.Render()
+			out = append(out, [2]*ast.Term{e.childKeys[hash], subtreeResult})
+		}
+		skipKeys := ast.NewSet(skipKeysList...)
+		// Copy over all keys that weren't deleted/modified.
+		x.Foreach(func(k, v *ast.Term) {
+			if skipKeys.Contains(k) {
+				return
+			}
+			out = append(out, [2]*ast.Term{k, v})
+		})
+		return ast.ObjectTerm(out...)
+	case ast.Set:
+		// Early exit if no modifications.
+		if len(e.childKeys) == 0 {
+			return e.value
+		}
+		// Build a new Set.
+		// Sets only can have deletions/new insertions, because the value
+		// *is* its own key.
+		skipKeysList := make([]*ast.Term, 0, len(e.childKeys))
+		out := make([]*ast.Term, 0, x.Len()+len(e.childKeys))
+		for hash, term := range e.childScalarValues {
+			skipKeysList = append(skipKeysList, e.childKeys[hash])
+			if term == nil {
+				continue // Delete case.
+			}
+			// Normal value case.
+			out = append(out, e.childScalarValues[hash])
+		}
+		// Only happens when this set hasn't been collapsed yet.
+		for hash, child := range e.childCompositeValues {
+			skipKeysList = append(skipKeysList, e.childKeys[hash])
+			if child == nil {
+				continue // Delete case.
+			}
+			// Normal value case.
+			subtreeResult := child.Render()
+			out = append(out, subtreeResult)
+		}
+		skipKeys := ast.NewSet(skipKeysList...)
+		// Copy over all keys that weren't deleted/modified.
+		x.Foreach(func(key *ast.Term) {
+			if skipKeys.Contains(key) {
+				return
+			}
+			out = append(out, key)
+		})
+		return ast.SetTerm(out...)
+	case *ast.Array:
+		// No early exit here, because we might have just deletes on the
+		// original array. We build a new Array with modified/deleted keys.
+		out := make([]*ast.Term, 0, e.insertions.Length())
+		eIdx := 0
+		for i := 0; i < e.insertions.Length(); i++ {
+			// If the index == 0, that indicates we should look up the next
+			// surviving original element.
+			// If the index == 1, that indicates we should look up that
+			// index in the child maps.
+			if e.insertions.Element(i) == 0 {
+				// Scan through the e.eliminated bit-vec, pick the
+				// first non-zero index. The then use that index to
+				// look up the element we should append from e.value.
+				foundIdx := false
+				for j := eIdx; j < e.eliminated.Length(); j++ {
+					if e.eliminated.Element(j) == 0 {
+						foundIdx = true
+						eIdx = j
+						break
+					}
+				}
+				if !foundIdx {
+					panic(fmt.Errorf("too many eliminated indexes in array, expected to find uneliminated index %d", i))
+				}
+				// Append element from original term.
+				out = append(out, x.Elem(eIdx))
+				eIdx++ // Bump the counter, so that we monotonically advance through the array.
+			} else {
+				// Append value from rendered child index.
+				if t, ok := e.childScalarValues[i]; ok {
+					out = append(out, t)
+				} else if child, ok := e.childCompositeValues[i]; ok {
+					t := child.Render()
+					// TODO: Nil check here?
```

Comment:
Haha, oh goodness. :sweat_smile: I had a bunch of those `// TODO: Nil check here?` comments copy/pasted around, but eliminated most of them because I saw `Render` was safe in virtually every instance.

In this particular case no nil check should be needed, because `nil` _should_ be an impossible result from `Render` here, assuming you're working on Arrays *only* through the `EditTree` APIs. I've added a comment to clarify this as part of my latest changes.

I am definitely open to the idea of programming more defensively here, but I'm also fine assuming that anyone going outside the public APIs should know what they're doing, and if they goof the bookkeeping and invariants, it's on them. I'll see about those invariants to the doc comments. :thinking: